### PR TITLE
Improve cards and general layout/styling

### PR DESCRIPTION
**Pull request by harindup**
- Change card layout
- Use proper local variable declaration keywords in changed files
- Reorder methods in classes so that public methods are before private ones
- Apply Justin's general layout to courses

### CODE REVIEW

File: frontend/src/components/pages/course-selection/CourseList.tsx [hunk 1]
```diff
@@ -1,75 +1,94 @@
...
-        )
+        });
     }
+
 }
 export default connect( () => ({}), {
-    selectCourse: (id: number) => push("courses/"+id),
-})(CourseList);
\ No newline at end of file
+    selectCourse: (id: number) => push(`courses/${id}`),
```
Comment (JustinPraas) [inline frontend/src/components/pages/course-selection/CourseList.tsx]: Is this a grave instead of a quote? How does this syntax work?

File: frontend/src/components/pages/course-selection/CourseList.tsx [hunk 2]
```diff
@@ -1,75 +1,94 @@
...
+        "#1976D2", "#0097A7", "#00796B", "#388E3C", "#AFB42B",
+        "#FBC02D", "#FFA000", "#F57C00", "#E64A19",
+    ];
+
+    render() {
+        if (this.props.courses.length === 0) {
+            return <Row />;
+        }
+        return (
+        <Row className="main-body-display px-2">
```
Comment (JustinPraas) [inline frontend/src/components/pages/course-selection/CourseList.tsx]: I believe this Row shouldn't be there. Looking at the page itself, you can see a scroll bar in the middle of the page. This is due to the fact that there are _two_ rows with _main-body-display_ classnames. One comes from here, the other comes from the render method of course_selection.

File: frontend/src/components/pages/course-selection/CourseList.tsx (hunk 1 above)
Comment (harindup) [reply frontend/src/components/pages/course-selection/CourseList.tsx]: Oh, yeah that enables you to use variable "parameters" inside the string. Pretty nifty to use instead of concatenation...

File: frontend/src/components/pages/course-selection/CourseList.tsx (hunk 2 above)
Comment (harindup) [reply frontend/src/components/pages/course-selection/CourseList.tsx]: Correct, thanks for spotting this!

File: frontend/src/styling/index.scss [hunk 3]
```diff
@@ -32,23 +32,37 @@ button {
     padding: 1em;
 }
 
-.course-selection-card {
-    height: 100px;
+.card {
```
Comment (JustinPraas) [inline frontend/src/styling/index.scss]: I'm also not sure how I feel about modifying reactstrap classes, since it will impact all elements using those classes (given they use the same stylesheet)

Comment (remco1112) [reply frontend/src/styling/index.scss]: I don't think that is a problem, in fact I think that is desirable. This ensures consistency through our application. If you have an exceptional case where you need a different styling you can always just overrule the styling with your own css class.

File: frontend/src/components/pages/course-selection/CourseSelection.tsx [hunk 4]
```diff
@@ -6,39 +6,98 @@ import { coursesRequestedAction, } from '../../../state/course-selection/action'
...
+    courses?: CourseDtoSummary[];
     requestCourses: () => {
-        type: String
-    }
+        type: string,
+    };
 }
 interface CourseSelectionState {
-    errorVisible: boolean,
+    errorVisible: boolean;
```
Comment (remco1112) [inline frontend/src/components/pages/course-selection/CourseSelection.tsx]: Why is this here? This is part of the state of the notification system.

File: frontend/src/components/pages/course-selection/CourseSelection.tsx [hunk 5]
```diff
@@ -6,39 +6,98 @@ import { coursesRequestedAction, } from '../../../state/course-selection/action'
...
-        if (this.props.courses.every((course: CourseDtoSummary) => course.role.name === API_STUDENT_ROLE)) {
-            return (
-                <Container>
-                    <CourseList mode={COURSE_LIST_ANY} courses={this.props.courses} />
-                </Container>
-            );
+        let lists = [];
+        const roles = new Set(this.props.courses.map((course: CourseDtoSummary) => course.role.name));
+
+        if (roles.size <= 1) {
```
Comment (remco1112) [inline frontend/src/components/pages/course-selection/CourseSelection.tsx]: I think we (I discussed this with Justin, Daan or probably both) agreed that the desired behavior is as follows: if you only have student courses, you just see "Courses", else you will see the role you have within that course. With this change, a teacher with only teacher courses will also see "Courses"...

File: frontend/src/components/pages/course-selection/CourseList.tsx [hunk 6]
```diff
@@ -1,75 +1,94 @@
...
+    ];
+
+    render() {
+        if (this.props.courses.length === 0) {
+            return <Row />;
+        }
+        return (
+        <Row>
+        { this.props.mode !== COURSE_LIST_ANY &&
+            <h3>
```
Comment (remco1112) [inline frontend/src/components/pages/course-selection/CourseList.tsx]: This looks very odd..

File: frontend/src/components/pages/course-selection/CourseSelection.tsx (hunk 4 above)
Comment (harindup) [reply frontend/src/components/pages/course-selection/CourseSelection.tsx]: I think this branch has stuff before the errors were merged to develop. I'll merge develop now and push again. What I did here was change the comma to a semicolon, as TypeScript interface fields should be.

File: frontend/src/components/pages/course-selection/CourseSelection.tsx (hunk 5 above)
Comment (harindup) [reply frontend/src/components/pages/course-selection/CourseSelection.tsx]: Oh, okay. I didn't know that, and thought it would be weird for teachers to see "Courses where you are a TA" and "Courses where you are a student", since they will never have anything under those headings.

File: frontend/src/components/pages/course-selection/CourseList.tsx (hunk 6 above)
Comment (harindup) [reply frontend/src/components/pages/course-selection/CourseList.tsx]: What looks odd?

File: frontend/src/components/pages/course-selection/CourseSelection.tsx (hunk 5 above)
Comment (remco1112) [reply frontend/src/components/pages/course-selection/CourseSelection.tsx]: They will only see those headings when there is actual courses in that category. So for a teacher, in practice they will only see "Courses where you are a teacher"

File: frontend/src/components/pages/course-selection/CourseList.tsx (hunk 6 above)
Comment (remco1112) [reply frontend/src/components/pages/course-selection/CourseList.tsx]: Apparantly I cannot comment within a change, it just shows you the whole change. The <h3> looks odd. Take an account that actually has different roles (or just create one) and you will see what I mean

File: frontend/src/components/pages/course-selection/CourseSelection.tsx (hunk 5 above)
Comment (harindup) [reply frontend/src/components/pages/course-selection/CourseSelection.tsx]: That wasn't the case earlier. It simply showed all three if at least one of the courses were non-student roles (see below). That's why I changed it to this. I can change it to something else.

screenshot 2019-02-19 at 19 06 30

File: frontend/src/styling/index.scss (hunk 3 above)
Comment (harindup) [reply frontend/src/styling/index.scss]: I thought the same as Remco. All cards in the application will now look like this, and I think it's good. I used a distinct class for the "coloured header" in the courses, and didn't change the bootstrap one.

File: frontend/src/components/pages/course-selection/CourseSelection.tsx (hunk 5 above)
Comment (remco1112) [reply frontend/src/components/pages/course-selection/CourseSelection.tsx]: CourseList renders nothing at all (so also no heading) when there is no courses to render

File: frontend/src/components/pages/course-selection/CourseList.tsx (hunk 6 above)
Comment (harindup) [reply frontend/src/components/pages/course-selection/CourseList.tsx]: Fixed

File: frontend/src/components/pages/course-selection/CourseSelection.tsx (hunk 4 above)
Comment (harindup) [reply frontend/src/components/pages/course-selection/CourseSelection.tsx]: Removed. I think you forgot to remove this earlier?
